src/gluonts/dataset/loader_v2.py

```diff
@@ -0,0 +1,189 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from typing import Iterable, Iterator, List, Callable, Optional
+import itertools
+import random
+from multiprocessing import Process, Queue, Event
+from queue import Empty
+
+import numpy as np
+
+from gluonts.dataset.common import DataBatch, Dataset
+from gluonts.dataset.util import MPWorkerInfo
+
+
+def stack(data, make_array_fn):
+    if isinstance(data[0], np.ndarray):
+        data = make_array_fn(data)
+    elif isinstance(data[0], (list, tuple)):
+        return list(stack(t, make_array_fn) for t in zip(*data))
+    return data
+
+
+def batchify(
+    data: List[dict], make_array_fn: Callable = np.asarray
+) -> DataBatch:
+    return {
+        key: stack(
+            data=[item[key] for item in data], make_array_fn=make_array_fn
+        )
+        for key in data[0].keys()
+    }
+
+
+class CyclicIterable(Iterable):
+    def __init__(self, base_iterable: Iterable) -> None:
+        self.base_iterable = base_iterable
+
+    def __iter__(self):
+        while True:
+            yield from self.base_iterable
+
+
+class PseudoShuffledIterable(Iterable):
+    def __init__(self, base_iterable: Iterable, buffer_length: int):
+        self.base_iterable = base_iterable
+        self.buffer_length = buffer_length
+        self.shuffle_buffer: list = []
+
+    def sample_from_buffer(self):
+        idx = random.randint(0, len(self.shuffle_buffer) - 1)
+        return self.shuffle_buffer.pop(idx)
+
+    def __iter__(self):
+        for x in self.base_iterable:
+            if len(self.shuffle_buffer) < self.buffer_length:
+                self.shuffle_buffer.append(x)
+            if len(self.shuffle_buffer) == self.buffer_length:
+                yield self.sample_from_buffer()
+        while len(self.shuffle_buffer):
+            yield self.sample_from_buffer()
+
+
+class MultiProcessIterator(Iterator):
+    def __init__(
+        self,
+        base_iterable: Iterable,
+        num_workers: int,
+        num_entries: int,
+        max_queue_size: Optional[int] = None,
+    ):
+        assert num_workers >= 1
+        assert num_entries >= 0
+        assert max_queue_size is None or max_queue_size >= num_workers
+
+        self.base_iterable = base_iterable
+        self.num_workers = num_workers
+        self.num_entries = num_entries
+        self.max_queue_size = (
+            max_queue_size if max_queue_size is not None else 5 * num_workers
+        )
+
+        self.data_queue: Queue = Queue(maxsize=self.max_queue_size)
```

Comment:
I think this is the wrong Queue type, if you are sharing the Queue with multiple processes. The correct one is afaik this one: https://github.com/awslabs/gluon-ts/blob/3774b1df6d2b9fb75fbabc181fadb188c34fc6db/src/gluonts/dataset/parallelized_loader.py#L660 i.e. a multiprocessing.manager().Queue()

Reply:
Yes, this seems to have solved the deadlock issues I was facing. Thanks!